crates/ruff_server/src/session/settings.rs

```diff
@@ -1,25 +1,452 @@
+/// during initialization.
+#[derive(Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq, Eq))]
+#[serde(untagged)]
+enum InitializationOptions {
+    #[serde(rename_all = "camelCase")]
+    HasWorkspaces {
+        global_settings: ClientSettings,
+        #[serde(rename = "settings")]
+        workspace_settings: Vec<WorkspaceSettings>,
+    },
+    GlobalOnly {
+        settings: Option<ClientSettings>,
+    },
+}
+
+impl AllSettings {
+    /// Initializes the controller from the serialized initialization options.
+    /// This fails if `options` are not valid initialization options.
+    pub(crate) fn from_value(options: serde_json::Value) -> Self {
+        Self::from_init_options(
+            serde_json::from_value(options)
+                .map_err(|err| {
+                    tracing::error!("Failed to deserialize initialization options: {err}. Falling back to default client settings...");
+                })
+                .unwrap_or_default(),
+        )
+    }
+
+    fn from_init_options(options: InitializationOptions) -> Self {
+        let (global_settings, workspace_settings) = match options {
+            InitializationOptions::GlobalOnly { settings } => (settings.unwrap_or_default(), None),
+            InitializationOptions::HasWorkspaces {
+                global_settings,
+                workspace_settings,
+            } => (global_settings, Some(workspace_settings)),
+        };
+
+        Self {
+            global_settings,
+            workspace_settings: workspace_settings
+                .into_iter()
+                .flatten()
+                .map(|workspace_settings| {
+                    (workspace_settings.workspace, workspace_settings.settings)
+                })
+                .collect(),
+        }
+    }
+}
+
+impl ResolvedClientSettings {
+    /// Resolves a series of client settings, prioritizing workspace settings over global settings.
+    /// Any fields not specified by either are set to their defaults.
+    pub(super) fn with_workspace(
+        workspace_settings: &ClientSettings,
+        global_settings: &ClientSettings,
+    ) -> Self {
+        Self::new_impl(&[workspace_settings, global_settings])
+    }
+
+    /// Resolves global settings only.
+    pub(super) fn global_only(global_settings: &ClientSettings) -> Self {
+        Self::new_impl(&[global_settings])
+    }
+
+    fn new_impl(all_settings: &[&ClientSettings]) -> Self {
         Self {
-            code_action_deferred_edit_resolution: code_action_data_support
-                && code_action_edit_resolution,
+            fix_all: Self::resolve_or(all_settings, |settings| settings.fix_all, true),
+            organize_imports: Self::resolve_or(
+                all_settings,
+                |settings| settings.organize_imports,
+                true,
+            ),
+            lint_enable: Self::resolve_or(
+                all_settings,
+                |settings| settings.lint.as_ref().and_then(|lint| lint.enable),
```

Comment:
Nit: It's possible to avoid some of the `and_then` by using the try operator (I'm okay with `and_then` if that's what you prefer)
```suggestion
                |settings| settings.lint.as_ref()?.enable,
```